File: CG_20250527163848/GeneratedCode.py
# Disclaimer: This output contains AI-generated content; user is advised to review it before consumption.
#*Start of AI Generated Content*

python
# ***********************************************************
# * Imports
# ***********************************************************

import os
import re
from flask import Flask, request, jsonify
from flask_pymongo import PyMongo
from werkzeug.security import generate_password_hash, check_password_hash
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
import smtplib

# ***********************************************************
# * Constants and Variables
# ***********************************************************

APP = Flask(__name__)

# Database Constants
DB_NAME = 'user_database'
DB_URL = 'mongodb://localhost:27017/'
APP.config['MONGO_URI'] = DB_URL + DB_NAME
MONGO = PyMongo(APP)

# Email Constants
EMAIL_ADDRESS = 'your-email@gmail.com'
EMAIL_PASSWORD = 'your-email-password'
SMTP_SERVER = 'smtp.gmail.com'
SMTP_PORT = 587

# Regex Patterns
EMAIL_PATTERN = r'^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$'
PHONE_PATTERN = r'^\d{3}-\d{3}-\d{4}$'
PASSWORD_PATTERN = r'^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*#?&])[A-Za-z\d@$!%*#?&]{8,}$'
NAME_PATTERN = r'^[a-zA-Z]{2,}$'
AGE_PATTERN = r'^\d{1,2}$'
SEX_PATTERN = r'^(Male|Female)$'
ADDRESS_PATTERN = r'^[a-zA-Z0-9\s]{5,}$'

# Food Preferences
FOOD_PREFERENCES = ['Indian', 'Chinese', 'French', 'Italian', 'Mexican', 'Japanese', 'Thai', 'American', 'Greek', 'Mediterranean']

# ***********************************************************
# * Functions
# ***********************************************************

def validate_email(email):
    """
    Validate email address.
    
    Args:
        email (str): Email address.
    
    Returns:
        bool: True if email is valid, False otherwise.
    """
    try:
        if re.match(EMAIL_PATTERN, email):
            return True
        else:
            raise ValueError('Invalid email address')
    except Exception as e:
        return str(e)

def validate_phone(phone):
    """
    Validate phone number.
    
    Args:
        phone (str): Phone number.
    
    Returns:
        bool: True if phone number is valid, False otherwise.
    """
    try:
        if re.match(PHONE_PATTERN, phone):
            return True
        else:
            raise ValueError('Invalid phone number')
    except Exception as e:
        return str(e)

def validate_password(password):
    """
    Validate password.
    
    Args:
        password (str): Password.
    
    Returns:
        bool: True if password is valid, False otherwise.
    """
    try:
        if re.match(PASSWORD_PATTERN, password):
            return True
        else:
            raise ValueError('Password must be at least 8 characters long and contain at least one uppercase letter, one lowercase letter, one number, and one special character')
    except Exception as e:
        return str(e)

def validate_name(name):
    """
    Validate name.
    
    Args:
        name (str): Name.
    
    Returns:
        bool: True if name is valid, False otherwise.
    """
    try:
        if re.match(NAME_PATTERN, name):
            return True
        else:
            raise ValueError('Invalid name')
    except Exception as e:
        return str(e)

def validate_age(age):
    """
    Validate age.
    
    Args:
        age (str): Age.
    
    Returns:
        bool: True if age is valid, False otherwise.
    """
    try:
        if re.match(AGE_PATTERN, age) and 18 <= int(age) <= 99:
            return True
        else:
            raise ValueError('Age must be between 18 and 99')
    except Exception as e:
        return str(e)

def validate_sex(sex):
    """
    Validate sex.
    
    Args:
        sex (str): Sex.
    
    Returns:
        bool: True if sex is valid, False otherwise.
    """
    try:
        if re.match(SEX_PATTERN, sex):
            return True
        else:
            raise ValueError('Invalid sex')
    except Exception as e:
        return str(e)

def validate_address(address):
    """
    Validate address.
    
    Args:
        address (str): Address.
    
    Returns:
        bool: True if address is valid, False otherwise.
    """
    try:
        if re.match(ADDRESS_PATTERN, address):
            return True
        else:
            raise ValueError('Invalid address')
    except Exception as e:
        return str(e)

def send_password_reset_link(email, password_reset_link):
    """
    Send password reset link to user's email.
    
    Args:
        email (str): User's email.
        password_reset_link (str): Password reset link.
    
    Returns:
        bool: True if email is sent successfully, False otherwise.
    """
    try:
        msg = MIMEMultipart()
        msg['From'] = EMAIL_ADDRESS
        msg['To'] = email
        msg['Subject'] = 'Password Reset Link'
        body = 'Please click on the following link to reset your password: ' + password_reset_link
        msg.attach(MIMEText(body, 'plain'))
        server = smtplib.SMTP(SMTP_SERVER, SMTP_PORT)
        server.starttls()
        server.login(EMAIL_ADDRESS, EMAIL_PASSWORD)
        text = msg.as_string()
        server.sendmail(EMAIL_ADDRESS, email, text)
        server.quit()
        return True
    except Exception as e:
        return str(e)

# ***********************************************************
# * API Endpoints
# ***********************************************************

@APP.route('/register', methods=['POST'])
def register():
    """
    Register a new user.
    
    Args:
        email (str): Email address or phone number.
        password (str): Password.
        confirm_password (str): Confirm password.
        security_question (str): Security question.
    
    Returns:
        dict: Registration status.
    """
    try:
        data = request.get_json()
        email = data['email']
        password = data['password']
        confirm_password = data['confirm_password']
        security_question = data['security_question']
        
        # Validate email or phone
        if '@' in email:
            email_validation = validate_email(email)
        else:
            email_validation = validate_phone(email)
        
        # Validate password
        password_validation = validate_password(password)
        
        # Check if password and confirm password match
        if password != confirm_password:
            return jsonify({'error': 'Password and confirm password do not match'}), 400
        
        # Check if any validation failed
        if email_validation != True or password_validation != True:
            return jsonify({'error': str(email_validation) if email_validation != True else str(password_validation)}), 400
        
        # Hash password
        hashed_password = generate_password_hash(password)
        
        # Insert user into database
        user = {
            'email': email,
            'password': hashed_password,
            'security_question': security_question
        }
        MONGO.db.users.insert_one(user)
        return jsonify({'message': 'User registered successfully'}), 201
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@APP.route('/forgot-password', methods=['POST'])
def forgot_password():
    """
    Send password reset link to user's email.
    
    Args:
        email (str): Email address.
        security_question_answer (str): Security question answer.
    
    Returns:
        dict: Password reset status.
    """
    try:
        data = request.get_json()
        email = data['email']
        security_question_answer = data['security_question_answer']
        
        # Find user in database
        user = MONGO.db.users.find_one({'email': email})
        
        # Check if user exists
        if user is None:
            return jsonify({'error': 'User not found'}), 404
        
        # Check if security question answer is correct
        if user['security_question'] != security_question_answer:
            return jsonify({'error': 'Incorrect security question answer'}), 400
        
        # Generate password reset link
        password_reset_link = 'https://example.com/reset-password/' + email
        
        # Send password reset link to user's email
        send_password_reset_link(email, password_reset_link)
        return jsonify({'message': 'Password reset link sent successfully'}), 200
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@APP.route('/save-personal-info', methods=['POST'])
def save_personal_info():
    """
    Save user's personal information.
    
    Args:
        email (str): Email address.
        name (str): Name.
        age (str): Age.
        sex (str): Sex.
        address (str): Address.
        profile_picture (str): Profile picture.
    
    Returns:
        dict: Personal info save status.
    """
    try:
        data = request.get_json()
        email = data['email']
        name = data['name']
        age = data['age']
        sex = data['sex']
        address = data['address']
        profile_picture = data['profile_picture']
        
        # Validate personal info
        name_validation = validate_name(name)
        age_validation = validate_age(age)
        sex_validation = validate_sex(sex)
        address_validation = validate_address(address)
        
        # Check if any validation failed
        if name_validation != True or age_validation != True or sex_validation != True or address_validation != True:
            return jsonify({'error': str(name_validation) if name_validation != True else str(age_validation) if age_validation != True else str(sex_validation) if sex_validation != True else str(address_validation)}), 400
        
        # Find user in database
        user = MONGO.db.users.find_one({'email': email})
        
        # Check if user exists
        if user is None:
            return jsonify({'error': 'User not found'}), 404
        
        # Update user's personal info in database
        MONGO.db.users.update_one({'email': email}, {'$set': {'name': name, 'age': age, 'sex': sex, 'address': address, 'profile_picture': profile_picture}})
        return jsonify({'message': 'Personal info saved successfully'}), 200
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@APP.route('/get-food-preferences', methods=['GET'])
def get_food_preferences():
    """
    Get list of food preferences.
    
    Returns:
        list: List of food preferences.
    """
    try:
        return jsonify({'food_preferences': FOOD_PREFERENCES}), 200
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@APP.route('/login', methods=['POST'])
def login():
    """
    Login a user.
    
    Args:
        email (str): Email address.
        password (str): Password.
    
    Returns:
        dict: Login status.
    """
    try:
        data = request.get_json()
        email = data['email']
        password = data['password']
        
        # Find user in database
        user = MONGO.db.users.find_one({'email': email})
        
        # Check if user exists
        if user is None:
            return jsonify({'error': 'User not found'}), 404
        
        # Check if password is correct
        if not check_password_hash(user['password'], password):
            return jsonify({'error': 'Incorrect email or password'}), 400
        
        return jsonify({'message': 'User logged in successfully'}), 200
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@APP.route('/logout', methods=['POST'])
def logout():
    """
    Logout a user.
    
    Returns:
        dict: Logout status.
    """
    try:
        return jsonify({'message': 'User logged out successfully'}), 200
    except Exception as e:
        return jsonify({'error': str(e)}), 500

if __name__ == '__main__':
    APP.run(debug=True)


#*End of AI Generated Content*

File: CG_20250527164302/GeneratedCode.java
// Disclaimer: This output contains AI-generated content; user is advised to review it before consumption.
//*Start of AI Generated Content*

java
// UserRegistrationAPI.java

import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.Filters;

import javax.validation.Valid;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

// Constants and Static Strings
public class Constants {
    public static final String EMAIL_REGEX = "^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$";
    public static final String PHONE_NUMBER_REGEX = "^\\d{3}-\\d{3}-\\d{4}$";
    public static final String PASSWORD_REGEX = "^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[@$!%*?&])[A-Za-z\\d@$!%*?&]{8,}$";
    public static final String NAME_REGEX = "^[a-zA-Z]{2,}$";
    public static final String AGE_RANGE = "^(1[8-9]|[2-9][0-9])$";
    public static final String SEX_OPTIONS = "^(Male|Female)$";
    public static final String ADDRESS_REGEX = "^[a-zA-Z0-9\\s]{5,}$";
    public static final String MONGO_URI = "mongodb://localhost:27017";
    public static final String DATABASE_NAME = "user_database";
    public static final String COLLECTION_NAME = "users";
    public static final String FOOD_PREFERENCES[] = {"Indian", "Chinese", "French", "Italian", "Mexican", "Japanese", "Thai", "American", "Greek", "Mediterranean"};
}

@RestController
public class UserRegistrationAPI {

    private MongoClient mongoClient;
    private MongoDatabase database;
    private MongoCollection<org.bson.Document> collection;

    public UserRegistrationAPI() {
        this.mongoClient = MongoClients.create(Constants.MONGO_URI);
        this.database = mongoClient.getDatabase(Constants.DATABASE_NAME);
        this.collection = database.getCollection(Constants.COLLECTION_NAME);
    }

    // Function Variables for Hardcoded Values
    private String getMongoURI() { return Constants.MONGO_URI; }
    private String getDatabaseName() { return Constants.DATABASE_NAME; }
    private String getCollectionName() { return Constants.COLLECTION_NAME; }
    private String[] getFoodPreferences() { return Constants.FOOD_PREFERENCES; }

    // User Registration
    @PostMapping("/register")
    public ResponseEntity<String> registerUser(@Valid @RequestBody UserRegistrationRequest request) {
        try {
            validateEmailOrPhone(request.getEmailOrPhone());
            validatePassword(request.getPassword(), request.getConfirmPassword());
            validateSecurityQuestion(request.getSecurityQuestion());

            org.bson.Document userDoc = new org.bson.Document("emailOrPhone", request.getEmailOrPhone())
                    .append("password", request.getPassword())
                    .append("securityQuestion", request.getSecurityQuestion());

            collection.insertOne(userDoc);
            return new ResponseEntity<>("User registered successfully", HttpStatus.CREATED);
        } catch (Exception e) {
            return new ResponseEntity<>(e.getMessage(), HttpStatus.BAD_REQUEST);
        }
    }

    // Validate Email or Phone
    private void validateEmailOrPhone(String emailOrPhone) throws Exception {
        if (Pattern.matches(Constants.EMAIL_REGEX, emailOrPhone)) {
            // Email validation successful
        } else if (Pattern.matches(Constants.PHONE_NUMBER_REGEX, emailOrPhone)) {
            // Phone number validation successful
        } else {
            throw new Exception("Invalid email address or phone number");
        }
    }

    // Validate Password
    private void validatePassword(String password, String confirmPassword) throws Exception {
        if (!Pattern.matches(Constants.PASSWORD_REGEX, password)) {
            throw new Exception("Password must be at least 8 characters long and contain at least one uppercase letter, one lowercase letter, one number, and one special character");
        }
        if (!password.equals(confirmPassword)) {
            throw new Exception("Passwords do not match");
        }
    }

    // Validate Security Question
    private void validateSecurityQuestion(String securityQuestion) throws Exception {
        if (securityQuestion.isEmpty()) {
            throw new Exception("Security question cannot be empty");
        }
    }

    // Forgot Password
    @PostMapping("/forgot-password")
    public ResponseEntity<String> forgotPassword(@Valid @RequestBody ForgotPasswordRequest request) {
        try {
            org.bson.Document userDoc = collection.find(Filters.eq("securityQuestion", request.getSecurityQuestion())).first();
            if (userDoc != null) {
                // Send password reset link to user's email or phone
                return new ResponseEntity<>("Password reset link sent successfully", HttpStatus.OK);
            } else {
                throw new Exception("Incorrect security question answer");
            }
        } catch (Exception e) {
            return new ResponseEntity<>(e.getMessage(), HttpStatus.BAD_REQUEST);
        }
    }

    // Save User Personal Information
    @PostMapping("/save-personal-info")
    public ResponseEntity<String> savePersonalInfo(@Valid @RequestBody PersonalInfoRequest request) {
        try {
            validateName(request.getName());
            validateAge(request.getAge());
            validateSex(request.getSex());
            validateAddress(request.getAddress());

            org.bson.Document userDoc = collection.find(Filters.eq("emailOrPhone", request.getEmailOrPhone())).first();
            if (userDoc != null) {
                userDoc.append("name", request.getName())
                        .append("age", request.getAge())
                        .append("sex", request.getSex())
                        .append("address", request.getAddress())
                        .append("profilePicture", request.getProfilePicture());

                collection.updateOne(Filters.eq("emailOrPhone", request.getEmailOrPhone()), new org.bson.Document("$set", userDoc));
                return new ResponseEntity<>("Personal information saved successfully", HttpStatus.OK);
            } else {
                throw new Exception("User not found");
            }
        } catch (Exception e) {
            return new ResponseEntity<>(e.getMessage(), HttpStatus.BAD_REQUEST);
        }
    }

    // Validate Name
    private void validateName(String name) throws Exception {
        if (!Pattern.matches(Constants.NAME_REGEX, name)) {
            throw new Exception("Invalid name");
        }
    }

    // Validate Age
    private void validateAge(String age) throws Exception {
        if (!Pattern.matches(Constants.AGE_RANGE, age)) {
            throw new Exception("Age must be between 18 and 99");
        }
    }

    // Validate Sex
    private void validateSex(String sex) throws Exception {
        if (!Pattern.matches(Constants.SEX_OPTIONS, sex)) {
            throw new Exception("Invalid sex");
        }
    }

    // Validate Address
    private void validateAddress(String address) throws Exception {
        if (!Pattern.matches(Constants.ADDRESS_REGEX, address)) {
            throw new Exception("Invalid address");
        }
    }

    // Get Food Preferences
    @PostMapping("/get-food-preferences")
    public ResponseEntity<String[]> getFoodPreferences() {
        return new ResponseEntity<>(getFoodPreferences(), HttpStatus.OK);
    }

    // Login
    @PostMapping("/login")
    public ResponseEntity<String> login(@Valid @RequestBody LoginRequest request) {
        try {
            org.bson.Document userDoc = collection.find(Filters.eq("emailOrPhone", request.getEmailOrPhone())).first();
            if (userDoc != null && userDoc.getString("password").equals(request.getPassword())) {
                return new ResponseEntity<>("Login successful", HttpStatus.OK);
            } else {
                throw new Exception("Incorrect email or password");
            }
        } catch (Exception e) {
            return new ResponseEntity<>(e.getMessage(), HttpStatus.BAD_REQUEST);
        }
    }

    // Logout
    @PostMapping("/logout")
    public ResponseEntity<String> logout(@Valid @RequestBody LogoutRequest request) {
        // Logout implementation
        return new ResponseEntity<>("Logout successful", HttpStatus.OK);
    }
}

// Request Classes
class UserRegistrationRequest {
    private String emailOrPhone;
    private String password;
    private String confirmPassword;
    private String securityQuestion;

    // Getters and Setters
    public String getEmailOrPhone() { return emailOrPhone; }
    public void setEmailOrPhone(String emailOrPhone) { this.emailOrPhone = emailOrPhone; }
    public String getPassword() { return password; }
    public void setPassword(String password) { this.password = password; }
    public String getConfirmPassword() { return confirmPassword; }
    public void setConfirmPassword(String confirmPassword) { this.confirmPassword = confirmPassword; }
    public String getSecurityQuestion() { return securityQuestion; }
    public void setSecurityQuestion(String securityQuestion) { this.securityQuestion = securityQuestion; }
}

class ForgotPasswordRequest {
    private String securityQuestion;

    // Getters and Setters
    public String getSecurityQuestion() { return securityQuestion; }
    public void setSecurityQuestion(String securityQuestion) { this.securityQuestion = securityQuestion; }
}

class PersonalInfoRequest {
    private String emailOrPhone;
    private String name;
    private String age;
    private String sex;
    private String address;
    private String profilePicture;

    // Getters and Setters
    public String getEmailOrPhone() { return emailOrPhone; }
    public void setEmailOrPhone(String emailOrPhone) { this.emailOrPhone = emailOrPhone; }
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    public String getAge() { return age; }
    public void setAge(String age) { this.age = age; }
    public String getSex() { return sex; }
    public void setSex(String sex) { this.sex = sex; }
    public String getAddress() { return address; }
    public void setAddress(String address) { this.address = address; }
    public String getProfilePicture() { return profilePicture; }
    public void setProfilePicture(String profilePicture) { this.profilePicture = profilePicture; }
}

class LoginRequest {
    private String emailOrPhone;
    private String password;

    // Getters and Setters
    public String getEmailOrPhone() { return emailOrPhone; }
    public void setEmailOrPhone(String emailOrPhone) { this.emailOrPhone = emailOrPhone; }
    public String getPassword() { return password; }
    public void setPassword(String password) { this.password = password; }
}

class LogoutRequest {
    // Logout request implementation
}


//*End of AI Generated Content*